### DISCUSSION TITLE
[Support] 使用sse返回的第一条数据会出现[object Object]

### DISCUSSION THREAD

**Discussion by fkomorebi** [midwayjs/midway#4174] [Q&A] [answered]
### Describe the problem（描述问题）

使用sse返回数据，但是返回的数据中意外地出现了[object Object]，这影响到了我的解析。虽然可以加以处理，但是会引起我的不安。
复现代码：
```javascript
import {
  Inject,
  Controller,
  Get,
  Query,
  HttpServerResponse,
} from '@midwayjs/core';
import { Context } from '@midwayjs/koa';
import { UserService } from '../service/user.service';

@Controller('/api')
export class APIController {
  @Inject()
  ctx: Context;

  @Inject()
  userService: UserService;

  @Get('/get_user')
  async getUser(@Query('uid') uid) {
    const res = new HttpServerResponse(this.ctx).sse();
    setTimeout(() => {
      res.send({
        data: 'test',
      });
      res.end();
    }, 3000);
    return res;
  }
}
```
返回的结果
![image](https://github.com/user-attachments/assets/c874d5ec-2236-4447-8103-093614a87857)

### Midway Versions（Midway 版本）

>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(not installed)
✓ @midwayjs/axios(not installed)
✓ @midwayjs/bootstrap(3.18.2)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/busboy(not installed)
✓ @midwayjs/cache-manager(not installed)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.18.0)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(not installed)
✓ @midwayjs/cross-domain(not installed)
✓ @midwayjs/decorator(not installed)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✓ @midwayjs/i18n(not installed)
✓ @midwayjs/info(3.18.2)
✓ @midwayjs/jwt(not installed)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.18.2)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(not installed)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(not installed)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(not installed)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/tags(not installed)
✓ @midwayjs/tenant(not installed)
✓ @midwayjs/typegoose(not installed)
✓ @midwayjs/typeorm(not installed)
✓ @midwayjs/upload(not installed)
✓ @midwayjs/validate(3.18.2)
✓ @midwayjs/version(not installed)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.18.2)
✓ @midwayjs/ws(not installed)
**********************************************************************
>> Check complete, all versions are healthy.
**********************************************************************

**Comment by waitingsong** [comment]
```ts
    const res = new HttpServerResponse(this.ctx).sse();
    setTimeout(() => {
      res.send({
        data: 'test',
      });
      res.end();
    }, 3000);
    return res;
```
 代码返回的 res 确是是个对象啊。setTimeout 应该改成 Promise 并返回这个 Promise

**Reply by fkomorebi** [reply]
谢谢您给的建议，对您给的建议我进行了如下尝试，此时返回的结果是 abc ，明显不符合预期。
```typescript
import {
  Inject,
  Controller,
  Get,
  Query,
  HttpServerResponse,
} from '@midwayjs/core';
import { Context } from '@midwayjs/koa';
import { UserService } from '../service/user.service';

@Controller('/api')
export class APIController {
  @Inject()
  ctx: Context;

  @Inject()
  userService: UserService;

  @Get('/get_user')
  async getUser(@Query('uid') uid) {
    const res = new HttpServerResponse(this.ctx).sse();
    return new Promise((resolve) => {
      setTimeout(() => {
        res.send({
          data: 'test',
        });
        res.end();
        resolve('abc');
      }, 3000);
    });
  }
}
```
从midway官方文档上看，是需要将该实例返回的。
![image](https://github.com/user-attachments/assets/e6e79e5d-fa10-4299-82ea-ac6194eb23e0)

**Comment by waitingsong** [comment]
try:
```ts

return new Promise((resolve) => {
      setTimeout(() => {
        res.send({
          data: 'test',
        });
        res.end();
        resolve(res); // <===
      }, 3000);
    });

```

**Reply by fkomorebi** [reply]
针对您给出的最近的建议我进行了更改，但是问题依旧存在：
![image](https://github.com/user-attachments/assets/d7111a00-ce8c-4e6f-98aa-e9c9febb5bc1)

**Comment by czy88840616** [comment]
请使用 EventSource 来接收数据，如果想打印在页面上，使用 `new HttpServerResponse(ctx).stream();`

**Reply by czy88840616** [reply]
sse 的测试 case

```ts
import { Controller, Get, HttpServerResponse } from '@midwayjs/core';

@Controller('/')
export class HomeController {
  @Get('/api/sse')
  async home(ctx) {
    const sse = new HttpServerResponse(ctx).stream();

    setInterval(() => {
      sse.send({
        data: 'test'
      });
    }, 1000);

    return sse;
  }
}

```

前端页面

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>SSE</title>
  <style>
    .message {
      padding: 8px;
      margin: 4px 0;
      border-bottom: 1px solid #eee;
    }
  </style>
  <script>
    document.addEventListener('DOMContentLoaded', () => {
      const sseDiv = document.getElementById('sse');
      
      // 创建 EventSource 实例连接到 SSE 端点
      const eventSource = new EventSource('/api/sse');
      
      // 监听服务器发送的消息
      eventSource.onmessage = function(event) {
        const messageDiv = document.createElement('div');
        messageDiv.className = 'message';
        messageDiv.textContent = event.data;
        sseDiv.appendChild(messageDiv);
        
        // 自动滚动到最新消息
        messageDiv.scrollIntoView({ behavior: 'smooth' });
      };
      
      // 错误处理
      eventSource.onerror = function(error) {
        console.error('SSE 连接错误:', error);
        eventSource.close();
      };
    });
  </script>
</head>
<body>
  <div id="sse"></div>
</body>
</html>
```

**Reply by fkomorebi** [reply]
您好！谢谢您的回答，我根据您给的方法进行测试。[object Object]还是会不可避免地出现在第一次回答中。我想这应该不是前端发送请求的方式和如何进行接收的问题。
![image](https://github.com/user-attachments/assets/68576c66-5715-48c2-bce9-b42d505e2847)

**Reply by czy88840616** [reply]
你没有明白，不能直接用网页（包括查看响应）来查看 SSE 的返回，你需要使用 "EventSource" 对象来获取。

**Reply by fkomorebi** [reply]
我想应该是**可以直接使用**网页（包括查看响应）来查看 SSE 的返回。如果使用 fetch 请求一个sse接口，其获取的是一个流式响应，获取的数据是原始的文本数据。只是相较于使用 EventSource，使用fetch的方式则需要自己解析。

我使用nest做了一个小demo进行测试，发现其返回的数据中没有带上不符合预期的字符。但是在midway中使用sse返回的第一条数据中带有 [object Object] ，这很意外，不清楚是 midway 的设计还是其他的问题。
```typescript
import { Controller, Get, Res, Sse } from '@nestjs/common';
import { AppService } from './app.service';
import { interval, map, Observable, take } from 'rxjs';

/**
 * AppController 负责处理 HTTP 请求和 SSE 事件流。
 */
@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}

  /**
   * 处理 SSE 请求，每秒向客户端发送一次 "Hello World" 消息。
   * @returns 一个 Observable 对象，包含 SSE 事件流。
   */
  @Sse('sse')
  sse(): Observable<MessageEvent> {
    return interval(1000).pipe(
      map(() => {
        const data = { event: 'tick', data: { hello: 'world' } };
        return new MessageEvent(data.event, {
          data: JSON.stringify(data.data),
        });
      }),
      take(5), // 只发送5个事件
    );
  }
}

```
![image](https://github.com/user-attachments/assets/b7d8deee-8721-4d98-86e6-69dea12d0818)

**Reply by czy88840616** [reply, marked as answer]
大概知道原因了，在第一次数据传输时，发送了一个{data: ' :ok'} 对象，这个逻辑参考了社区库的实现，有点问题，正确的应该拿到

```
: ok
data: test

data: test
```

才对

**Reply by czy88840616** [reply]
#4176 不太影响业务，因为 EventSource 会忽略不符合 SSE 格式规范的数据行。

**Reply by fkomorebi** [reply]
谢谢您为我解答！可算是知道了问题所在。
